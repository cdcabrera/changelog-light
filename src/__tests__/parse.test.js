// spell-checker: disable
const {
  getCommitType,
  formatChangelogMessage,
  parseCommitMessage,
  parseCommits,
  semverBump,
  getComparisonCommitHashes
} = require('../parse');

describe('Parse', () => {
  it('should return commit types', () => {
    expect(getCommitType()).toMatchSnapshot('getCommitType');
    expect(getCommitType({ isAllowNonConventionalCommits: true }).general).toMatchSnapshot('non-conventional-commits');
  });

  it('should get the first, last commits', () => {
    const commitLog = `
      LAST1f12345b597123453031234555bvvvccacee refactor(file): lorem updates (#8)
      53a12345479ef91123456e921234548ac4123450 feat(dolor): issues/20 sit enhancements (#8)
      d1234537b5e94a6512345xeb96503312345x18d2 fix(build): eslint, jsdoc updates (#16)
      e5c456ea12345vv4610fa4aff7812345ss31b1e2 chore(build): npm packages (#15)
      FIRST12345dd312345d42123123131231ca11235 Initial-like commit
    `;

    const comparisonObjNoReleaseCommit = getComparisonCommitHashes({
      getGit: () => commitLog,
      getReleaseCommit: () => ''
    });
    expect(comparisonObjNoReleaseCommit).toMatchSnapshot('first and last, no release commit');

    const comparisonObjReleaseCommit = getComparisonCommitHashes({
      getGit: () => `${commitLog}\nREALFIRST123452345d42123123131231ca11235 chore(release): 0.1.0`,
      getReleaseCommit: () => 'REALFIRST123452345d42123123131231ca11235'
    });
    expect(comparisonObjReleaseCommit).toMatchSnapshot('first and last, release commit');
  });

  it('should parse a commit message', () => {
    const commitMessageRefactor = '1f12345b597123453031234555b6d25574ccacee refactor(file): lorem updates (#8)';
    const commitMessageFeature = '1f12345b597123453031234555b6d25574ccacee feat(dolor): issues/20 sit enhancements';
    const commitMessageNonCC = '1f12345b597123453031234555b6d25574ccacee Initial commit';

    expect({
      refactor: parseCommitMessage(commitMessageRefactor),
      feat: parseCommitMessage(commitMessageFeature),
      general: parseCommitMessage(commitMessageNonCC)
    }).toMatchSnapshot('parseCommitMessages');
  });

  it('should format a changelog commit message', () => {
    const commitMessageRefactor = '1f12345b597123453031234555b6d25574ccacee refactor(file): lorem updates (#8)';
    const commitMessageFeature = '1f12345b597123453031234555b6d25574ccacee feat(dolor): issues/20 sit enhancements';
    const commitMessageNonCC = '1f12345b597123453031234555b6d25574ccacee Initial commit';

    expect({
      refactor: formatChangelogMessage(parseCommitMessage(commitMessageRefactor), {
        getRemoteUrls: () => ({
          commitUrl: 'https://localhost/lorem/ipsum/commitsmock/',
          prUrl: 'https://localhost/lorem/ipsum/prmock/'
        })
      }),
      feat: formatChangelogMessage(parseCommitMessage(commitMessageFeature)),
      general: formatChangelogMessage(parseCommitMessage(commitMessageNonCC))
    }).toMatchSnapshot('formatChangelogMessages');
  });

  it('should parse a commit listing using conventional commit types and semver', () => {
    const commitLog = `
      1f12345b597123453031234555b6d25574ccacee refactor(file): lorem updates (#8)
      53a12345479ef91123456e921234548ac4123450 feat(dolor): issues/20 sit enhancements (#8)
      611234511234543c39c1234536dc01234521549c fix(build): npm packages (#18)
      fe7d312345xe604d8328d025612345925123457b build(deps): bump codecov/codecov-action from 1.1.0 to 1.1.1 (#19)
      d1234537b5e94a6512345xeb96503312345x18d2 fix(build): eslint, jsdoc updates (#16)
      e5c456ea12345vv4610fa4aff7812345ss31b1e2 chore(build): npm packages (#15)
      6f0f55d32412345221234512345d2345c3212345 chore(build): npm packages (#15)
      112345ca212345c8d2d123450c31234588f12345 build(deps): bump actions/cache from 1 to 2 (#12)
      512345d6712345a1234581234501234516b12345 build(deps): bump actions/checkout from 1 to 2 (#11)
      12345d71e12345d2fc712345dd411234586b850a build(deps): aggregated checks, updates (#10)
      1412345dd312345d4212345d53e12345dca12345 build(deps): bump actions/github-script from 5 to 6 (#9)
      12345dd312345d421231231312312345dca11235 Initial commit
    `;

    const commitObj = parseCommits(undefined, { getGit: () => commitLog });
    const urlPathObj = parseCommits(
      { commitPath: 'sit', prPath: 'dolor', remoteUrl: 'https://localhost/lorem/ipsum' },
      { getGit: () => commitLog }
    );
    const generalCommitsObj = parseCommits(
      { remoteUrl: 'https://localhost/lorem/ipsum' },
      { getGit: () => commitLog, isAllowNonConventionalCommits: true }
    );

    expect({
      default: parseCommits(),
      commits: commitObj,
      urls: urlPathObj,
      generalCommits: generalCommitsObj
    }).toMatchSnapshot('parsed commits');
    expect({ default: semverBump(), commits: semverBump(commitObj) }).toMatchSnapshot('semver bump');
  });
});
