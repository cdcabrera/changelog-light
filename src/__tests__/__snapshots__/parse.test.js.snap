// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse should format a changelog commit message: formatChangelogMessages 1`] = `
{
  "feat": "* **dolor** issues/20 sit enhancements  (1f12345)",
  "general": "*  Initial commit  (1f12345)",
  "refactor": "* **file** lorem updates  ([#8](https://localhost/lorem/ipsum/prmock/8)) ([1f12345](https://localhost/lorem/ipsum/commitsmock/1f12345b597123453031234555b6d25574ccacee))",
}
`;

exports[`Parse should get the first, last commits: first and last, no release commit 1`] = `
{
  "first": null,
  "last": null,
}
`;

exports[`Parse should get the first, last commits: first and last, release commit 1`] = `
{
  "first": "REALFIRST123452345d42123123131231ca11235",
  "last": "LAST1f12345b597123453031234555bvvvccacee",
}
`;

exports[`Parse should parse a commit listing using conventional commit types and semver: parsed commits 1`] = `
{
  "commits": {
    "build": {
      "commits": [
        "* **deps** bump codecov/codecov-action from 1.1.0 to 1.1.1  (#19) (fe7d312)",
        "* **deps** bump actions/cache from 1 to 2  (#12) (112345c)",
        "* **deps** bump actions/checkout from 1 to 2  (#11) (512345d)",
        "* **deps** aggregated checks, updates  (#10) (12345d7)",
        "* **deps** bump actions/github-script from 5 to 6  (#9) (1412345)",
      ],
      "description": "Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)",
      "title": "Builds",
      "value": "build",
    },
    "chore": {
      "commits": [
        "* **build** npm packages  (#15) (e5c456e)",
        "* **build** npm packages  (#15) (6f0f55d)",
      ],
      "description": "Other changes that don't modify src or test files",
      "title": "Chores",
      "value": "chore",
    },
    "feat": {
      "commits": [
        "* **dolor** issues/20 sit enhancements  (#8) (53a1234)",
      ],
      "description": "A new feature",
      "title": "Features",
      "value": "feat",
    },
    "fix": {
      "commits": [
        "* **build** npm packages  (#18) (6112345)",
        "* **build** eslint, jsdoc updates  (#16) (d123453)",
      ],
      "description": "A bug fix",
      "title": "Bug Fixes",
      "value": "fix",
    },
    "refactor": {
      "commits": [
        "* **file** lorem updates  (#8) (1f12345)",
      ],
      "description": "A code change that neither fixes a bug nor adds a feature",
      "title": "Code Refactoring",
      "value": "refactor",
    },
  },
  "default": {},
  "generalCommits": {
    "build": {
      "commits": [
        "* **deps** bump codecov/codecov-action from 1.1.0 to 1.1.1  (#19) (fe7d312)",
        "* **deps** bump actions/cache from 1 to 2  (#12) (112345c)",
        "* **deps** bump actions/checkout from 1 to 2  (#11) (512345d)",
        "* **deps** aggregated checks, updates  (#10) (12345d7)",
        "* **deps** bump actions/github-script from 5 to 6  (#9) (1412345)",
      ],
      "description": "Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)",
      "title": "Builds",
      "value": "build",
    },
    "chore": {
      "commits": [
        "* **build** npm packages  (#15) (e5c456e)",
        "* **build** npm packages  (#15) (6f0f55d)",
      ],
      "description": "Other changes that don't modify src or test files",
      "title": "Chores",
      "value": "chore",
    },
    "feat": {
      "commits": [
        "* **dolor** issues/20 sit enhancements  (#8) (53a1234)",
      ],
      "description": "A new feature",
      "title": "Features",
      "value": "feat",
    },
    "fix": {
      "commits": [
        "* **build** npm packages  (#18) (6112345)",
        "* **build** eslint, jsdoc updates  (#16) (d123453)",
      ],
      "description": "A bug fix",
      "title": "Bug Fixes",
      "value": "fix",
    },
    "general": {
      "commits": [
        "*  Initial commit  (12345dd)",
      ],
      "description": "Commits without category",
      "title": "General",
      "value": "general",
    },
    "refactor": {
      "commits": [
        "* **file** lorem updates  (#8) (1f12345)",
      ],
      "description": "A code change that neither fixes a bug nor adds a feature",
      "title": "Code Refactoring",
      "value": "refactor",
    },
  },
  "urls": {
    "build": {
      "commits": [
        "* **deps** bump codecov/codecov-action from 1.1.0 to 1.1.1  ([#19](https://localhost/lorem/ipsum/dolor/19)) ([fe7d312](https://localhost/lorem/ipsum/sit/fe7d312345xe604d8328d025612345925123457b))",
        "* **deps** bump actions/cache from 1 to 2  ([#12](https://localhost/lorem/ipsum/dolor/12)) ([112345c](https://localhost/lorem/ipsum/sit/112345ca212345c8d2d123450c31234588f12345))",
        "* **deps** bump actions/checkout from 1 to 2  ([#11](https://localhost/lorem/ipsum/dolor/11)) ([512345d](https://localhost/lorem/ipsum/sit/512345d6712345a1234581234501234516b12345))",
        "* **deps** aggregated checks, updates  ([#10](https://localhost/lorem/ipsum/dolor/10)) ([12345d7](https://localhost/lorem/ipsum/sit/12345d71e12345d2fc712345dd411234586b850a))",
        "* **deps** bump actions/github-script from 5 to 6  ([#9](https://localhost/lorem/ipsum/dolor/9)) ([1412345](https://localhost/lorem/ipsum/sit/1412345dd312345d4212345d53e12345dca12345))",
      ],
      "description": "Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)",
      "title": "Builds",
      "value": "build",
    },
    "chore": {
      "commits": [
        "* **build** npm packages  ([#15](https://localhost/lorem/ipsum/dolor/15)) ([e5c456e](https://localhost/lorem/ipsum/sit/e5c456ea12345vv4610fa4aff7812345ss31b1e2))",
        "* **build** npm packages  ([#15](https://localhost/lorem/ipsum/dolor/15)) ([6f0f55d](https://localhost/lorem/ipsum/sit/6f0f55d32412345221234512345d2345c3212345))",
      ],
      "description": "Other changes that don't modify src or test files",
      "title": "Chores",
      "value": "chore",
    },
    "feat": {
      "commits": [
        "* **dolor** issues/20 sit enhancements  ([#8](https://localhost/lorem/ipsum/dolor/8)) ([53a1234](https://localhost/lorem/ipsum/sit/53a12345479ef91123456e921234548ac4123450))",
      ],
      "description": "A new feature",
      "title": "Features",
      "value": "feat",
    },
    "fix": {
      "commits": [
        "* **build** npm packages  ([#18](https://localhost/lorem/ipsum/dolor/18)) ([6112345](https://localhost/lorem/ipsum/sit/611234511234543c39c1234536dc01234521549c))",
        "* **build** eslint, jsdoc updates  ([#16](https://localhost/lorem/ipsum/dolor/16)) ([d123453](https://localhost/lorem/ipsum/sit/d1234537b5e94a6512345xeb96503312345x18d2))",
      ],
      "description": "A bug fix",
      "title": "Bug Fixes",
      "value": "fix",
    },
    "refactor": {
      "commits": [
        "* **file** lorem updates  ([#8](https://localhost/lorem/ipsum/dolor/8)) ([1f12345](https://localhost/lorem/ipsum/sit/1f12345b597123453031234555b6d25574ccacee))",
      ],
      "description": "A code change that neither fixes a bug nor adds a feature",
      "title": "Code Refactoring",
      "value": "refactor",
    },
  },
}
`;

exports[`Parse should parse a commit listing using conventional commit types and semver: semver bump 1`] = `
{
  "commits": {
    "bump": "minor",
    "weight": 11.899999999999999,
  },
  "default": {
    "bump": "patch",
    "weight": 0,
  },
}
`;

exports[`Parse should parse a commit message: parseCommitMessages 1`] = `
{
  "feat": {
    "description": "issues/20 sit enhancements",
    "hash": "1f12345b597123453031234555b6d25574ccacee",
    "prNumber": "",
    "scope": "dolor",
    "type": "feat",
    "typeScope": "feat(dolor)",
  },
  "general": {
    "description": "Initial commit",
    "hash": "1f12345b597123453031234555b6d25574ccacee",
    "prNumber": "",
    "scope": undefined,
    "type": "general",
    "typeScope": undefined,
  },
  "refactor": {
    "description": "lorem updates ",
    "hash": "1f12345b597123453031234555b6d25574ccacee",
    "prNumber": "8",
    "scope": "file",
    "type": "refactor",
    "typeScope": "refactor(file)",
  },
}
`;

exports[`Parse should return commit types: getCommitType 1`] = `
{
  "build": {
    "description": "Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)",
    "title": "Builds",
    "value": "build",
  },
  "chore": {
    "description": "Other changes that don't modify src or test files",
    "title": "Chores",
    "value": "chore",
  },
  "ci": {
    "description": "Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)",
    "title": "Continuous Integrations",
    "value": "ci",
  },
  "docs": {
    "description": "Documentation only changes",
    "title": "Documentation",
    "value": "docs",
  },
  "feat": {
    "description": "A new feature",
    "title": "Features",
    "value": "feat",
  },
  "fix": {
    "description": "A bug fix",
    "title": "Bug Fixes",
    "value": "fix",
  },
  "perf": {
    "description": "A code change that improves performance",
    "title": "Performance Improvements",
    "value": "perf",
  },
  "refactor": {
    "description": "A code change that neither fixes a bug nor adds a feature",
    "title": "Code Refactoring",
    "value": "refactor",
  },
  "revert": {
    "description": "Reverts a previous commit",
    "title": "Reverts",
    "value": "revert",
  },
  "style": {
    "description": "Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)",
    "title": "Styles",
    "value": "style",
  },
  "test": {
    "description": "Adding missing tests or correcting existing tests",
    "title": "Tests",
    "value": "test",
  },
}
`;

exports[`Parse should return commit types: non-conventional-commits 1`] = `
{
  "description": "Commits without category",
  "title": "General",
  "value": "general",
}
`;
